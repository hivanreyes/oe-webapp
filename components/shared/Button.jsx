import React from 'react'
import PropTypes from 'prop-types'
import cx from 'classnames'
import style from './styles/button.scss'

export const TYPE = {
  PRIMARY: 'PRIMARY',
  SECONDARY: 'SECONDARY',
}

export const SIZE = {
  BIG: 'BIG',
  SMALL: 'SMALL',
}

const getButtonClassNames = ({ buttonType, size }) =>
  cx({
    [style.button]: true,
    [style.primary]: buttonType === TYPE.PRIMARY,
    [style.secondary]: buttonType === TYPE.SECONDARY,
    [style.big]: size === SIZE.BIG,
    [style.small]: size === SIZE.SMALL,
  })

const Button = props => {
  const { label, onClick } = props
  const buttonClassNames = getButtonClassNames(props)
  return (
    <button onClick={onClick} className={buttonClassNames}>
      {label}
    </button>
  )
}

Button.defaultProps = {
  buttonType: TYPE.PRIMARY,
  size: SIZE.SMALL,
}

Button.propTypes = {
  label: PropTypes.string.isRequired,
  buttonType: PropTypes.oneOf([TYPE.PRIMARY, TYPE.SECONDARY]),
  size: PropTypes.oneOf([SIZE.SMALL, SIZE.BIG]),
  onClick: PropTypes.func,
}

export default Button
