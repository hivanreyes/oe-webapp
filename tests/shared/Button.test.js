import React from 'react'
import { shallow } from 'enzyme'
import Button, { TYPE, SIZE } from '../../components/shared/Button'

describe('<Button />', () => {
  it('should shallow without errors', () => {
    const btn = shallow(<Button label={'Follow'} />)
    expect(btn).toMatchSnapshot()
  })
  it('should render label as its child', () => {
    const btn = shallow(<Button label={'Follow'} />)
    expect(btn.props().children).toBe('Follow')
  })
  it('should have `button primary small` as default className', () => {
    const btn = shallow(<Button label={'Follow'} />)
    expect(btn.props().className).toBe('button primary small')
  })
  it('should change its className depending on buttonType and size', () => {
    const prim = shallow(<Button label={'Follow'} buttonType={TYPE.PRIMARY} />)
    const sec = shallow(<Button label={'Follow'} buttonType={TYPE.SECONDARY} />)
    const small = shallow(<Button label={'Follow'} size={SIZE.SMALL} />)
    const big = shallow(<Button label={'Follow'} size={SIZE.BIG} />)
    expect(prim.props().className).toBe('button primary small')
    expect(sec.props().className).toBe('button secondary small')
    expect(small.props().className).toBe('button primary small')
    expect(big.props().className).toBe('button primary big')
  })
})
