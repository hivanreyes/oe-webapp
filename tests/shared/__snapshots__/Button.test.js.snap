// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should mount without errors 1`] = `
ReactWrapper {
  "length": 1,
  Symbol(enzyme.__unrendered__): <Button
    buttonType="PRIMARY"
    label="Follow"
    size="SMALL"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "buttonType": "PRIMARY",
      "label": "Follow",
      "size": "SMALL",
    },
    "ref": null,
    "rendered": Object {
      "instance": <button
        class="button primary small"
      >
        Follow
      </button>,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Follow",
        "className": "button primary small",
      },
      "ref": null,
      "rendered": Array [
        "Follow",
      ],
      "type": "button",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "buttonType": "PRIMARY",
        "label": "Follow",
        "size": "SMALL",
      },
      "ref": null,
      "rendered": Object {
        "instance": <button
          class="button primary small"
        >
          Follow
        </button>,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Follow",
          "className": "button primary small",
        },
        "ref": null,
        "rendered": Array [
          "Follow",
        ],
        "type": "button",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`<Button /> should render without errors 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "button primary small",
    },
    "children": Array [
      Object {
        "data": "Follow",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "button",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`<Button /> should shallow without errors 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Button
    buttonType="PRIMARY"
    label="Follow"
    size="SMALL"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "Follow",
      "className": "button primary small",
    },
    "ref": null,
    "rendered": "Follow",
    "type": "button",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Follow",
        "className": "button primary small",
      },
      "ref": null,
      "rendered": "Follow",
      "type": "button",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
